subprojects/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGeneratorTest.groovy

```diff
@@ -0,0 +1,76 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.instrumentation.processor.codegen.jvmbytecode
+
+import com.google.testing.compile.Compilation
+import org.gradle.internal.instrumentation.InstrumentationCodeGenTest
+
+import static com.google.testing.compile.CompilationSubject.assertThat
+
+class InterceptJvmCallsGeneratorTest extends InstrumentationCodeGenTest {
+
+    def "should generate interceptors for inherited calls"() {
+        given:
+        def givenSource = source """
+            package org.gradle.test;
+            import org.gradle.internal.instrumentation.api.annotations.*;
+            import org.gradle.internal.instrumentation.api.annotations.CallableKind.*;
+            import org.gradle.internal.instrumentation.api.annotations.ParameterKind.*;
+            import org.gradle.internal.classpath.declarations.*;
+            import java.io.File;
+
+            @SpecificJvmCallInterceptors(generatedClassName = InterceptorDeclaration.JVM_BYTECODE_GENERATED_CLASS_NAME)
+            public class FileInterceptorsDeclaration {
+                @InterceptCalls
+                @InstanceMethod
+                @InterceptInherited
+                public static File[] intercept_listFiles(@Receiver File thisFile) {
+                    return new File[0];
+                }
+            }
+        """
+
+        when:
+        Compilation compilation = compile(givenSource)
+
+        then:
+        def expectedJvmInterceptors = source """
+            package org.gradle.internal.classpath;
+            class InterceptorDeclaration_JvmBytecodeImpl extends MethodVisitorScope implements JvmBytecodeCallInterceptor {
+                @Override
+                public boolean visitMethodInsn(String className, int opcode, String owner, String name,
+                        String descriptor, boolean isInterface, InstrumentationMetadata metadata) {
+                    if (metadata.isInstanceOf(owner, "java/io/File")) {
+                        /**
+                         * Intercepting instance method: {@link java.io.File#listFiles()}
+                         * Intercepted by {@link FileInterceptorsDeclaration#intercept_listFiles(File)}
+                        */
+                        if (name.equals("listFiles") && descriptor.equals("()[Ljava/io/File;") && opcode == Opcodes.INVOKEVIRTUAL) {
+                            _INVOKESTATIC(FILE_INTERCEPTORS_DECLARATION_TYPE, "intercept_listFiles", "(Ljava/io/File;)[Ljava/io/File;");
+                            return true;
+                        }
+                    }
+                    return false;
+                }
+            }
+        """
```

Comment:
🤔 It feels both brittle to expect the exact text to be generated in this test. It is also somewhat error-prone in that it is easy to insert the same error to the expected text that we have in the implementation.

I wonder if there's a better way to test this. Can we test this in some higher-level test instead, where we can actually run the generated code and observe its effect?

Reply:
We actually don't test exact text, full class has also imports, a constructor and fields. With `containsElementsIn` we compare just part of AST. So text spaces and even comments don't play any role here.

We are testing that we get the expected generated class. I think this is good for some basic tests that we generate what we want. We could also write "functional test", by loading classes or smth. like that, but I think we cover most of that with integration tests.